model/riscv_insts_vext_utils.sail

```diff
@@ -0,0 +1,703 @@
+  let end_element   : int                       = get_end_element();
+  let tail_ag       : agtype                    = get_vtype_vta();
+  let mask_ag       : agtype                    = get_vtype_vma();
+  mask_helper       : vector('n, dec, bool)     = undefined;
+  result            : vector('n, dec, bits('m)) = undefined;
+
+  if start_element < 0 then {
+    /* start element is not valid */
+    result      = undefined;
+    mask_helper = undefined;
+  }
+  else {
+    /* Determine the actual number of elements when lmul < 1 */
+    var real_num_elem : int = undefined;
+    if lmul >= 1.0 then {
+      real_num_elem = num_elem;
+    }
+    else {
+      real_num_elem = floor(lmul * to_real(num_elem))
+    };
+    assert(num_elem >= real_num_elem);
+
+    foreach (i from 0 to (num_elem - 1)) {
+      if i < start_element then {
+        /* Prestart elements defined by vstart */
+        result[i] = vd_val[i];
+        mask_helper[i] = false
+      }
+      else if i > end_element then {
+        /* Tail elements defined by vl */
+        if tail_ag == UNDISTURBED then {
+          result[i] = vd_val[i];
+        }
+        else if tail_ag == AGNOSTIC then {
+          //result[i] = sail_sign_extend(0b1, vsew_bits)
+          result[i] = vd_val[i];
+        };
+        mask_helper[i] = false
+      }
+      else if i >= real_num_elem then {
+        /* Tail elements defined by lmul < 1 */
+        if tail_ag == UNDISTURBED then {
+          result[i] = vd_val[i];
+        }
+        else if tail_ag == AGNOSTIC then {
+          //result[i] = sail_sign_extend(0b1, vsew_bits)
+          result[i] = vd_val[i];
+        };
+        mask_helper[i] = false
+      }
+
+      else if vm_val[i] == false then {
+        /* Inactive body elements defined by vm */
+        if mask_ag == UNDISTURBED then {
+          result[i] = vd_val[i]
+        }
+        else if mask_ag == AGNOSTIC then {
+          //result[i] = sail_sign_extend(0b1, vsew_bits)
+          result[i] = vd_val[i]
+        };
+        mask_helper[i] = false
+      }
+      else {
+        /* Active body elements */
+        mask_helper[i] = true;
+      }
+    };    
+  };
+
+  (result, mask_helper)
+}
+
+/* Mask handling for carry functions that use masks as input/output */
+/* Only prestart and tail elements are masked in a mask value */
+val init_masked_result_carry : forall 'n 'm, 8 <= 'm <= 128. (int('n), int('m), real, vector('n, dec, bool)) -> (vector('n, dec, bool), vector('n, dec, bool)) effect {escape, rreg, undef, wreg}
+function init_masked_result_carry(num_elem, vsew_bits, lmul, vd_val) = {
+  let start_element : int                   = get_start_element();
+  let end_element   : int                   = get_end_element();
+  mask_helper       : vector('n, dec, bool) = undefined;
+  result            : vector('n, dec, bool) = undefined;
+
+  /* Determine the actual number of elements when lmul < 1 */
+  var real_num_elem : int = undefined;
```

Comment:
This doesn't need to be mutable, use an if expression (pattern repeated in later functions)

Reply:
Fixed